Fix vsphere detach bug

gnufied opened the pull request:
Fixes Kubernetes not detaching volumes from shutdown nodes even after pods that are using the volume has been deleted. Fixes https://github.com/kubernetes/kubernetes/issues/67900


This is a variant of https://github.com/kubernetes/kubernetes/pull/63413


/sig storage

```release-note
Fix volume detaches from powered-off vSphere VMs
```

w-leads committed:
Fix bug in detaching disks in vsphere cp

VCP fails to detach disk from powered-off node which is still in
vcenter inventory. That's caused because VCP tries to detach disks
after nodemanager has unregistered the node information.

Use configmap as a fallback when node is not available

gnufied commented:
/assign @SandeepPissay

gnufied commented:
/sig vmware

gnufied commented:
/test pull-kubernetes-verify

abrarshivani reviewed (commented):
Deletion of entries from configmap is missing. Do you plan to handle it in this PR?
Rest looks good to me.

gnufied commented:
/test pull-kubernetes-integration

gnufied commented:
@abrarshivani alright, so I have tested this with a multinode cluster deployed using plain k8s on ESXi 6.7 and it does indeed appear to work and volumes are being detached without problem.

I am not handling removing of entries from configmap though. I think, that is something that should be done when node is terminated and removed from vcenter inventory too.

nikopen commented:
@gnufied is this fix only needed in 1.11? any issue linked with this PR (apart from #67825 )?

gnufied commented:
@nikopen this fix is needed for 1.9,1.10 and 1.11 (or whichever k8s versions are still considered supported except master). The PR you linked(https://github.com/kubernetes/kubernetes/pull/67825) just disables multiattach issue with vSphere, the underlying problem still is - we are NOT detaching volumes from shutdown nodes at all in these versions even after pods that were running on shutdown nodes has been deleted.

nikopen commented:
Indeed 67825 simply circumvented around the issue.

Did something else fix this in 1.12 onwards?

/lgtm

nikopen commented:
/unassign

abrarshivani commented:
@gnufied Thanks for testing. 
/lgtm
/approve

gnufied commented:
@abrarshivani can you please lgtm again? I have refactored the configmap cache to its own file and updated the code to ensure that - we do not cross limit of configmap data size. Also added tests for it. Underlying logic is still the same. Also I have tested the PR in multinode cluster and it works as expected.

gnufied commented:
/kind bug

nikopen commented:
/lgtm

@gnufied why isn't this needed in 1.12 or later? something else fixed it?

gnufied commented:
@nikopen we don't need this in 1.12 because we are not deleting node api objects when a node is shutdown and hence we don't need to save node's UUID in a configmap etc.

abrarshivani commented on pkg/cloudprovider/providers/vsphere/nodeuuidcache.go:
```diff
@@ -0,0 +1,184 @@
+)
+
+const (
+	// Since node name can be of length 253 and uuid is of length 36
```
+1

abrarshivani commented on pkg/cloudprovider/providers/vsphere/nodeuuidcache.go:
```diff
@@ -0,0 +1,184 @@
+func (cache *nodeUUIDCache) pruneCache() {
+	for k, v := range cache.nodeUUIDMap {
+		// if node is not deleted then we can still fetch node's UUID from node object itself
+		if !v.deleted {
```
When a node is deleted NodeDeleted method is called where the latest node object is passed. This object has field VolumeAttached which can be used to know if volumes are still attached to the node. With this information, nodes with no volume attached can be deleted first.

gnufied replied:
Discussed this offline, I will file a separate issue for handling more than 3600 nodes in more robust way.

abrarshivani commented on pkg/cloudprovider/providers/vsphere/nodeuuidcache.go:
```diff
@@ -0,0 +1,184 @@
+		return fmt.Errorf("error loading uuid in configmap %s with %v", cache.configMapName, err)
+	}
+	cache.nodeUUIDMap[nodeName] = &nodeUUID{uuid: uuid}
+	configMap.Data = cache.getUUIData()
```
I would suggest, to persist information that nodes that are marked as deleted in configmap. Since if kube-controller-manager restarts, vSphere Cloud Provider can keep the nodes which are marked as deleted if there is a need to prune the configmap.

abrarshivani commented:
/approve

liggitt reviewed (commented):
If you want to specify the role/binding in a yaml file for vsphere deployers to install, that seems more reasonable than setting this up on all clusters, even non-vsphere ones

gnufied commented:
@liggitt just so I understand this correctly - your recommendation is provide an YAML file with necessary roles and bindings and after that it is the job of installation tool to actually apply that YAML. Is there any other example of this?

liggitt commented:
Similar to the yaml manifests for default storage class when running in a particular cloud provider env: 

https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/storage-class/aws/default.yaml

liggitt commented:
> @nikopen we don't need this in 1.12 because we are not deleting node api objects when a node is shutdown and hence we don't need to save node's UUID in a configmap etc.

backporting that change to 1.11 seems like a more reasonable approach than this

gnufied commented:
> > @nikopen we don't need this in 1.12 because we are not deleting node api objects when a node is shutdown and hence we don't need to save node's UUID in a configmap etc.
> 
> backporting that change to 1.11 seems like a more reasonable approach than this

I am not comfortable with changing lifecycle of node api objects in a minor release that has to be backported all the way back to 1.9.  We may have customer scripts/utilities etc that depends on it. Also - we don't know what breaks in kubernetes itself if we do that. Having said that, I am not super familiar with vsphere code and I will have to rely on @SandeepPissay @abrarshivani to verify if that is something worth considering.

gnufied committed:
Provider cluster addon for vsphere rbac policies

Install cloudprovider specific rbac policies

gnufied commented:
@liggitt moved the rbac policy installation to a cluster add-on. It is currently only installed via `./hack/local-cluster-up.sh`  since `./kube-up.sh` does not support vsphere. PTAL.

gnufied commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -489,6 +489,13 @@ func ClusterRoles() []rbacv1.ClusterRole {
 			},
 		},
+		{
+			ObjectMeta: metav1.ObjectMeta{Name: "system:vsphere-cloud-provider"},
```
reduced the scope to just the namespace. It is no longer crated as a clusterole.

gnufied commented:
/test pull-kubernetes-e2e-gce

gnufied commented:
/test pull-kubernetes-e2e-kops-aws

gnufied commented:
@SandeepPissay yes in releases prior to 1.12, node object is being deleted when a node is shutdown and that is why I proposed this PR only for 1.9,1.10 and 1.11 releases.  

What @liggitt is saying is - instead of this PR (which uses configmap cache for storing node name/uuid mapping) - can we backport the PR that changed the node deletion behavior on shutdown?

nikopen commented:
FYI 1.9 is not active anymore - it's just 1.10 and 1.11 as per 3 active versions

nikopen commented:
IMO at this point since it's about bugfixing for previous releases, if backporting the 1.12 PR has the same effect (fixes the bug) then it's a viable option. Do you expect any different behaviour with that, or extra PRs needed as well?

SandeepPissay commented:
Backporting the 1.12 PR is a better solution. Do you know which PR fixes the node removal bug?

gnufied commented:
The PR that changed lifecycle of node object for vsphere nodes is - https://github.com/kubernetes/kubernetes/pull/66619  but that PR alone isn't enough to fix the bug in question because it relies on a new cloud interface function `InstanceShutdownByProviderID` which was introduced by a different PR - https://github.com/kubernetes/kubernetes/pull/59323 (which exists in 1.11 but not in 1.10).

It is possible to rework a patch that just changes shutdown behaviour for VMs and isn't a backport of 1.12 PR. Another thing to keep in mind is, changing lifecycle of node objects(i.e keeping them around in case of shutdown) won't fix the detach issue. Volumes still will not be detached. But then user will have option of force deleting pods for forcing detach of volumes from powered-down nodes.  So basically, vSphere behaviour will become same as GCE.

nikopen commented:
@gnufied so if #66619 is backported to 1.10 + 1.11 and #59323 to 1.10, it should be good to go?

gnufied commented:
@nikopen I have opened a new smaller PR - https://github.com/kubernetes/kubernetes/pull/70291 which just changes behaviour of shutdown nodes. if we backport it, then at least if user force deletes his pods volumes will be detached.

gnufied commented:
Since we went with different solution. Closing
/close
